src/bson_mapper/bson_archiver.hpp

```diff
@@ -220,20 +234,37 @@ class BSONOutputArchive : public cereal::OutputArchive<BSONOutputArchive> {
         switch (_nodeTypeStack.top()) {
             case OutputNodeType::StartArray:
                 _bsonBuilder.open_array();
+                ++_arrayNestingLevel;
             // Don't break so that this array can also be closed.
             case OutputNodeType::InArray:
                 _bsonBuilder.close_array();
+                --_arrayNestingLevel;
                 break;
             case OutputNodeType::StartObject:
-                if (_nodeTypeStack.size() > 1 || _objAsRootElement) {
-                    _bsonBuilder.open_document();
+                // Only consider creating a new document if we're not in dot notation mode, or if we
+                // are within an array somewhere.
+                if (!_dotNotationMode || _arrayNestingLevel > 0) {
+                    if (_nodeTypeStack.size() > 1 || _objAsRootElement) {
+                        _bsonBuilder.open_document();
+                    }
+                } else if (_nodeTypeStack.size() > 1) {
+                    // Push back a dummy name so we don't accidentally pop an empty stack when
+                    // closing this empty document.
+                    _embeddedNameStack.push_back("");
                 }
             // Don't break so that this document can also be closed.
             case OutputNodeType::InObject:
-                if (_nodeTypeStack.size() > 1) {
-                    _bsonBuilder.close_document();
-                } else if (_objAsRootElement) {
-                    _bsonBuilder.close_document();
+                // Only close the document if we are not in dot notation mode or if we are within an
+                // array somewhere.
+                if (!_dotNotationMode || _arrayNestingLevel > 0) {
+                    if (_nodeTypeStack.size() > 1) {
+                        _bsonBuilder.close_document();
+                    } else if (_objAsRootElement) {
+                        _bsonBuilder.close_document();
+                    }
+                } else if (_nodeTypeStack.size() > 1) {
+                    // Pop the name of this embedded document off the stack.
+                    _embeddedNameStack.pop_back();
```

Comment:
Ah, it's caught on line 259. Great, no need to add more checks here – that's what I had missed.